exe-unit/src/network/inet.rs

```diff
@@ -498,36 +633,128 @@ impl Proxy {
                     }
                 };
 
+                let handle = get_handle(&network, &conn.meta);
+
                 log::debug!(
-                    "[inet] proxy conn: forward received bytes ({} B) to {:?}",
-                    vec.len(),
-                    conn
+                    "[inet] proxy conn: forward received bytes ({} B) to {conn:?}, {handle:?}",
+                    vec.len()
                 );
 
+                // After packet will processed in network stack,
+                // it will be received in `inet_egress_handler` function.
                 if let Err(e) = network.send(vec, conn).await {
                     log::debug!("[inet] proxy conn: send error: {}", e);
                 };
             }
 
-            let _ = proxy.unbind(desc).await;
+            let _ = proxy.disconnect(handle).await;
             log::debug!("[inet] proxy conn closed: {:?}", desc);
         }
         .right_future())
     }
 
     async fn unbind(&self, desc: SocketDesc) -> Result<()> {
-        log::debug!("[inet] proxy unbind: {:?}", desc);
+        log::debug!("[inet] proxy unbind: {desc:?}");
 
         let meta = ConnectionMeta::try_from(desc)?;
         let key = (&meta).proxy_key()?;
-        let mut inner = self.state.write().await;
 
-        if let Some(mut conn) = inner.remotes.remove(&key) {
-            let _ = conn.close().await;
+        if let Some(mut conn) = { self.state.write().await.remotes.remove(&key) } {
+            log::trace!("Closing channel for: {desc:?}");
+            let _ = conn.sender.close().await;
+            conn.abort.abort();
         }
-
         Ok(())
     }
+
+    /// Disconnect inside smoltcp stack, but don't clear our structures.
+    /// It will be done later in unbind function, after stack will send
+    /// IngressEvent::Disconnected event.
+    async fn disconnect(&self, socket: SocketHandle) -> Result<()> {
+        log::debug!("[inet] proxy socket disconnect: {socket}");
+
+        let network = { self.state.read().await.network.clone() };
+        Ok(network.stack.disconnect(socket).await?)
```

Comment:
```suggestion
        network.stack.disconnect(socket).await
```
Btw `disconnect` can never return an error so that should be changed in `ya-relay` to return `()` instead of `Result<()>`.

Reply:
Can't use the suggested change, the `Result` types have different errors. And while I see that the other library internally never returns `Err(..)`, I don't want to introduce a dependency on that behavior here.